transhumandesign/kag-base
# [added] Despawn Flickering to some blobs

## Status

- **READY**: this PR is (to the best of your knowledge) ready to be incorporated into the game.

## Description

This PR makes it so some blobs that are about to despawn will start to flicker.
The flickering becomes faster the shorter the remaining time is.
If you pick up the blob, the decay time is reset, so there is no "flickering item in hands" and the player is rewarded for picking up an item that is about to expire.

Fixes #1385.

----

```
[added] DespawnFlicker.as - this script causes blobs to flicker visually in the last seconds before despawn.
		This script has been added to:
		Heart, Grain, Steak, Log, Fishy, mat_arrows, mat_ballistabolts, mat_ballistabombbolts, mat_bombs,
		mat_waterbombs, mat_bombarrows, mat_firearrows, mat_waterarrows.
[added] DecayUnused.as to Heart, Log and Fishy - this means those blobs do not despawn if kept in inventory or when attached to player or catapult.
[changed] Grain uses DecayUnused.as instead of TimeOut.as, with a decay time of 60 seconds set to it (same time as was set in TimeOut.as).
[changed] Steak uses DecayUnused.as instead of TimeOut.as, with a decay time of 60 seconds set to it (same time as was set in TimeOut.as).
[changed] Steak will gib if it is destroyed via damage, instead of nothing happening
[changed] Steak will not gib if it despawns via server_SetTimeToDie().
[changed] Fishy will gib if it is destroyed via damage and if its age is greater than 0, instead of nothing happening
[changed] Fishy will not gib if it despawns via server_SetTimeToDie().
[changed] Log will not gib if it despawns via server_SetTimeToDie().
[removed] TimeOut.as - as it did the same thing as DecayUnused.as.
[fixed] Dead Fishy will not gain size.
```
----

These blobs will not flicker as I decided against it for now:
- Corpses
- Flying or collided arrow
- Flying ballista bolt

Materials Wood, Stone and Gold don't have `DecayUnused.as` and `DespawnFlickering.as` added to them since they are never set to despawn at once, but in steps in quantity.

Tested offline and online and no bugs observed.

## Steps to Test or Reproduce

Go to offline or online Sandbox.
`!log`, `!heart`, `!grain`, `!steak`, `!mat_arrows`, `!mat_bombs` etc.
or `!fishy` and have the fishy die.
Then wait until the blob despawns.
Notice it will be flickering in the last 4 seconds.

`!heart`
Pick up the heart, keep it in your hand or place it in an inventory and come again later.
Notice the heart will not have disappeared.

`!steak`
Keep hitting the steak by pickaxe or sword jab.
Notice it will gib now.

`!fishy`
Kill the fishy and keep hitting it.
Notice it will gib now.

`!fishy`
Kill the fishy and wait for a while.
Notice it will not grow to larger size while dead now.
